README.md

```diff
@@ -1,16 +1,148 @@
+intended to be the first open-source application performance management (APM) library available 
+in all the main programming languages including C/C++, Java, Go, Ruby, PHP, Python, C#, Node.js,
+Objective-C and Erlang.
 
-## Stats Design
+## Ecosystem Design
 
-## Supported Encodings
+![alt text][EcosystemLayers]
+
+### Layers
+
+#### Service Exporters
+Each backend service MUST implement this API to export data to their services.
+
+#### OpenCensus Library
+This is what we design/describe in the current document.
+
+#### Manually instrumented Frameworks
+We are going to instrument some of the most popular frameworks for each language using the 
+OpenCensus library to allow users to get traces/stats when they use these frameworks.
+
+#### Tools for automatic instrumentation
+Some of the languages may support libraries for automatic instrumentation. Java for example can use
+byte-code manipulation (monkey patching) to provide an agent that automatically instruments a 
+binary. Note: not all the languages support this.
+
+#### Application
+This is customer's application/binary.
+
+## Library Design
+
+### Namespace and Package
+* For details about the library package names structure see [Namespace and Package][NamespaceAndPackage].
+
+### Components
+This section focuses on the important components that each OpenCensus library must have to 
+support all the functionalities.
+
+Here is a layering structure of the proposed OpenCensus library:
+
+![alt text][LibraryComponents]
+
+
+#### Generic Context
+Some of the features like tracing (distributed tracing) and tagging (possibly others) need a way 
+to propagate a specific context (trace, tags) in-process (possibly between threads) and function
+calls.
+
+The key elements of the Generic Context Support are:
+* Every implementation MUST offer an explicit or implicit generic Context propagation mechanism 
+that allows different sub-contexts to be propagated.
+* Languages that already have this support, like [Go][goContext] or C# (ExecutionContext), MUST 
+use the language supported generic context instead of building their own.
+* For an explicit generic context implementation you can look at the Java [io.grpc.Context][gRPCContext].
+
+
+#### Trace
+Trace component is designed to support distributed tracing (see [dapper paper][DapperPaper]). This 
+component allows collecting and recording in memory trace data, tracks active requests, and keeps
+local samples for interesting requests.
+
+The key elements of the API can be broken down as:
+* A Span represents a single operation within a trace. Spans can be nested to form a trace tree. 
+* Libraries must allow users to record tracing events for a span (attributes, annotations, links, 
+etc.).
+* Spans are carried in the Generic Context. Libraries MUST provide a way of getting, manipulating,
+and replacing the Span in the current context.
+* Libraries MUST provide a means of dynamically controlling the trace global config at runtime (e
```

Comment:
Is this truly a MUST? I don't think it will be feasible for PHP to implement this.

Reply:
:+1: I'd reword it as: *Libraries SHOULD provide a means of dynamically controlling the trace global config at runtime (immediately or between requests).

Reply:
I think I misunderstood this line. It must be configurable at runtime, but not necessarily changable while running.

Reply:
Why is it hard to do it at runtime (change)? I relaxed it to SHOULD but I would like to understand.